### PR TITLE
V2 documentation updates

### PR DESCRIPTION
**Pull request by Cawllec**
## Goal

- Adds yardoc strings to all steps and publicly accessible methods and classes
- Adds yard and appropriate rake task to construct documentation using yard
- Adds (currently disabled) prep work for build and publishing documentation to github pages as part of CI

## TODO
- Enable and test Github pages CI integration

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -204,13 +258,6 @@ This request will match:
 { "fruit": { "apple": "some nonsense" } }
 ```
 
-# On dealing with multiple requests
```
Comment (tomlongridge) [inline README.md]: Is this no longer relevant?

File: README.md [hunk 2]
```diff
@@ -223,32 +270,6 @@ bundle exec bugsnag-maze-runner features/something.feature
 Add the `--verbose` option to print script output and a trace of what Ruby file
 is being run.
 
-## Logging data on scenario failures
```
Comment (tomlongridge) [inline README.md]: Sections L226-251 deliberately removed?

File: lib/features/steps/request_assertion_steps.rb [hunk 3]
```diff
@@ -54,102 +69,180 @@ def read_key_path hash, key_path
...
+#
+# @step_input parameter_name [String] The parameter to test
 Then("the {string} query parameter is not null") do |parameter_name|
   assert_not_nil(parse_querystring(Server.current_request)[parameter_name][0], "The '#{parameter_name}' query parameter should not be null")
 end
+
+# Tests that a query parameter is a timestamp.
+#   Uses the regex /^\d{4}\-\d{2}\-\d{2}T\d{2}:\d{2}:[\d\.]+Z?$/
+#
+# @step_input parameter_naem [String] The parameter to test
```
Comment (tomlongridge) [inline lib/features/steps/request_assertion_steps.rb]:
```suggestion
# @step_input parameter_name [String] The parameter to test
```

File: lib/features/steps/request_assertion_steps.rb [hunk 4]
```diff
@@ -54,102 +69,180 @@ def read_key_path hash, key_path
...
 Then("the {string} query parameter is a timestamp") do |parameter_name|
   param = parse_querystring(Server.current_request)[parameter_name][0]
   assert_match(TIMESTAMP_REGEX, param)
 end
 
-## MULTIPART REQUEST ASSERTIONS
+# Tests the number of fields a multipart request contains.
+#
+# @step_input part_count [Integer] The number of expected feilds
```
Comment (tomlongridge) [inline lib/features/steps/request_assertion_steps.rb]:
```suggestion
# @step_input part_count [Integer] The number of expected fields
```

File: README.md (hunk 1 above)
Comment (Cawllec) [reply README.md]: No, this functionality has been removed in v2.  Multiple requests should now be handled using steps from https://github.com/bugsnag/maze-runner/pull/66/files#diff-04c6e90faac2675aa89e2176d2eec7d8R194

File: README.md (hunk 2 above)
Comment (Cawllec) [reply README.md]: We no-longer handle failures this way, although I believe we do have some sort of failure mechanism so I'll have a look into that.

File: Rakefile [hunk 5]
```diff
@@ -17,8 +17,28 @@ namespace :test do
...
+  end
+
+  task :publish do
+    throw StandardError("Docs have not been generated") unless Dir.exist?("doc")
+    Dir.chdir("doc") do
+      `git init`
+      `git checkout -b gh-pages`
+      `git add .`
+      # `git commit -m "Docs update"`
+      # `git push -f git@github.com:bugsnag/maze-runner.git gh-pages`
```
Comment (Cawllec) [inline Rakefile]: I'm being cautious on this command, as I don't want to accidentally overwrite non-"gh-pages" branches.  Can someone verify this for me?

File: README.md (hunk 2 above)
Comment (Cawllec) [reply README.md]: I've added information about how logging works now in the `troubleshooting` section.

**Pull request merged**